openedx/features/badging/models.py

```diff
@@ -18,3 +24,89 @@ class Badge(models.Model):
 
     def __unicode__(self):
         return self.name
+
+    @classmethod
+    def get_remaining_badges(cls, user_id, community_id, community_type):
+        badges_in_community = UserBadge.objects.filter(user_id=user_id,
+                                                       community_id=community_id,
+                                                       badge_id__type=community_type)
+
+        if badges_in_community:
+            latest_earned = badges_in_community.latest('date_earned')
+            latest_threshold = Badge.objects.get(pk=latest_earned.badge_id, type=community_type).threshold
+            remaining_badges = Badge.objects.filter(type=community_type) \
+                                            .exclude(threshold__lte=latest_threshold) \
+                                            .order_by('threshold')
+        else:
+            remaining_badges = Badge.objects.filter(type=community_type).order_by('threshold')
+        
+        remaining_badges_json = {}
+        for badge in remaining_badges:
+            remaining_badges_json[badge.id] = {'name': badge.name,
+                                               'description': badge.description,
+                                               'threshold': badge.threshold,
+                                               'type': badge.type,
+                                               'image': badge.image}
+        return remaining_badges_json
+
+
+class UserBadge(models.Model):
+    """
+        This model represents what badges are assigned to which users in
+        communities (both course and team communities)
+
+        Each object of this model represents the assignment of one badge (specified
+        by the `badge_id` foreign key) to a certain user (specified by the `user`
+        foreign key) in a `community`. Each `community` is related to a course
+        which is specified by the `course_id`. A certain badge is only awarded once
+        in a community, the unique_together constraint in Meta class makes sure
+        that there are no duplicate objects in the model.
+    """
+    user = models.ForeignKey(User, db_index=True, on_delete=models.CASCADE)
```

Comment:
Let's please make sure the use case: `Delete all badges of a user if user is deleted`.

Reply:
Noted.